usplash_write.8
.TH usplash_write "8" "February 2007" Ubuntu
.\"
.SH NAME
usplash_write \- send commands to a running usplash daemon
.\"
.SH SYNOPSIS
\fBusplash_write\fR \fICOMMAND\fR...
.\"
.SH DESCRIPTION
Sends the given commands to a running usplash daemon.
If no daemon is running,
.B usplash_write
returns without doing anything.
.\"
.SH COMMANDS
.TP
.BI QUIT
Instructs the daemon to terminate, restoring the console that was
previously active.
.\"
.TP
.BI TIMEOUT " secs"
Changes the number of seconds to wait for a command before exiting to
that given.
.\"If zero is given it will wait forever.
.\"
.TP
.BI CLEAR
Clears any text currently displayed.
.\"
.TP
.BI TEXT " text"
Displays the given text in the area on the screen reserved for that,
scrolling any existing text up.
.\"
.TP
.BI TEXT\-URGENT " text"
Displays the given text in the same way as TEXT above even if messages would
normally be hidden due to the verbosity setting in effect.
.\"
.TP
.BI STATUS " status"
Displays the given status string to the right of the last line of text
in the same colour.
.\"
.TP
.BI SUCCESS " status"
Displays the given status string to the right of the last line of text
in a colour used to indicate success.
.\"
.TP
.BI FAILURE " status"
Displays the given status string to the right of the last line of text
in a colour used to indicate failure.
.\"
.TP
.BI PROGRESS " percentage"
Update the progress bar to show that the given percentage (0-100) of the
process has been completed.
.\"
.TP
.BI PULSATE
Set the progress bar to display a throbber (an indicator bouncing from left
to right and back again).
.\"
.TP
.BI INPUT " prompt"
Displays the given prompt and waits for one line of user input.
The line can then be read from /dev/.initramfs/usplash_outfifo within the
timeout period.
.\"
.TP
.BI INPUTQUIET " prompt"
Works in the same way as INPUT above with the exception that the user input
is not echoed to the console.
Useful for password prompts.
.\"
.TP
.BI INPUTENTER " prompt"
Works in the same way as INPUT above with the exception that nothing is
echoed to the console.
Useful for "Press enter to reboot".
.\"
.TP
.BI INPUTTIMEOUT " timeout prompt"
Works in the same way as INPUT above, but returns the empty string
after the given number of deciseconds. 
.\"
.TP
.BI INPUTCHAR
Checks for a single pending key press and writes the character to
/dev/.initramfs/usplash_outfifo. If no key was pressed, this will be
an empty string. Useful for polling for input (such as "Press ESC to
skip") while doing something in the background (such as running fsck).
.\"
.TP
.BI VERBOSE " mode"
Changes verbosity at runtime. Possible values for
.I mode
are
.B on
or
.B true
for verbose mode,
.B off
or
.B false
for quiet mode,
or
.B default
for the default mode when usplash started.

.\"
.SH EXIT STATUS
.B usplash_write
always returns an exit status of zero by default.

However, if the environment variable
.B FAIL_NO_USPLASH
is set, then failure to open the usplash fifo (which usually means
that usplash is not running) will exit with 1. This can be tested
before reading something from the FIFO (which would wait forever).

.\"
.SH AUTHOR
Written by Matthew Garrett <mjg59@srcf.ucam.org>
.\"
.SH REPORTING BUGS
Report bugs at
.br
http://launchpad.net/distros/ubuntu/+source/usplash/+filebug
.\"
.SH COPYRIGHT
Copyright \(co 2006 Canonical Ltd.
.br
Copyright \(co 2005 Matthew Garrett <mjg59@srcf.ucam.org>

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.SH SEE ALSO
.BR usplash (8)


debian/changelog
usplash (0.5.19-1) unstable; urgency=low

  * New upstream release (closes: #442881)

 -- maximilian attems <maks@debian.org>  Sun, 04 May 2008 22:46:05 +0200

usplash (0.5.8-3) unstable; urgency=low

  * Add back lsb-base-logging.sh since it's need to show boot progress
  * Revert the conflicts with sysvinit since we are compatible with
    previous sysvinit

 -- Otavio Salvador <otavio@ossystems.com.br>  Sat, 29 Dec 2007 21:45:51 -0200

usplash (0.5.8-2) experimental; urgency=low

  [ Otavio Salvador ]
  * Use the new sysvinit API for splash handling and drop the lsb based
    one.
  * Use the now supported Vcs-{Git,Browser} fields.
  * Bump standards-version to 3.7.3 (no changes needed).
  * Conflicts with sysvinit (<< 2.86.ds1-39) since we now use the
    provided API.

  [ Petter Reinholdtsen ]
  * splash-functions: testing fixups

  [ maximilian attems ]
  * usplash.init: Call console-screen.sh with start arg (closes: #457655)
  * splash-functions: small coding style fixups.
  * usplash_write.8: Comment out unimplemented TIMEOUT feature.
  * debian/rules: kill lpia arch mismerge addition.

 -- Otavio Salvador <otavio@ossystems.com.br>  Thu, 27 Dec 2007 21:41:53 -0200

usplash (0.5.8-1) unstable; urgency=low

  * Ubuntu sync, minus lpia and build-dep. (closes: #445885, #453796, #453847)
  * lsb-base-logging.sh: Fix unbound access. (closes: #432948)
    Thanks "James Youngman" <jay@gnu.org> for patch.

 -- maximilian attems <maks@debian.org>  Thu, 06 Dec 2007 22:33:44 +0100

usplash (0.5.2-3) unstable; urgency=low

  * Add XS-Vcs-{Git,Browser} fields and drop XS-Vcs-Bzr one.
  * Add a watch file to make easier to follow Ubuntu updates.
  * Do not check for kernel param on usplash_quit method. Thanks to Geert
    Stappers <stappers@stappers.nl> for the patch. (closes: #429621, #422920)
  * Restrict the architecture building to: amd64, i386, powerpc, sparc

 -- Otavio Salvador <otavio@ossystems.com.br>  Sun, 08 Jul 2007 19:59:06 -0300

usplash (0.5.2-2) unstable; urgency=low

  * Use usplash_write TEXT on lsb_{sucess,failure,warning}
    methods. Thanks to "David Härdeman" <david@hardeman.nu>
    for the patch. (closes: #423118, #385969)
  * Fix usplash_down logic. Thanks to "David Härdeman" <david@hardeman.nu>
    for the patch. (closes: #423098)
  * Restore console when timing out. Thanks to "David Härdeman"
    <david@hardeman.nu> for the patch. (closes: #387640)
  * Fix link to libusplash.so.0 to avoid dpkg-shlibdeps confusion. Thanks
    to Michael Biebl <biebl@debian.org> for the patch. (closes: #424731)
  * Don't ignore clean errors.

 -- Otavio Salvador <otavio@ossystems.com.br>  Wed, 27 Jun 2007 11:21:11 -0300

usplash (0.5.2-1) unstable; urgency=low

  * New upstream release.
  * debian/control:
    - Change libgd2-dev build-dependency to libgd2-noxpm-dev;
    - Change libusplash0 dependency to make it bin-NMU friendly;
    - Remove a duplicate entry of section field at libusplash-dev;
  * debian/lsb-base-logging-ubuntu.sh: Add a hack while we don't migrate
    to the proposed splash methods to avoid to patch KDM and GDM.

 -- Otavio Salvador <otavio@ossystems.com.br>  Sun, 17 Jun 2007 18:02:06 -0300

usplash (0.4-43-3) unstable; urgency=low

  * debian/control: Fix section of libusplash-dev.
    Mirror active Maintainers and Uploaders from usplash-theme-debian.
    Replaces and Provides bootsplash. (closes: #422761)
  * lsb-base-logging.sh: Fix log_action_end_msg return code to be consistent
    with lsb-base. Thanks martin f krafft <madduck@debian.org> for patch.
    (closes: #422624)
  * initramfs/scripts/init-top/usplash: Disable splash too on single bootarg.

 -- maximilian attems <maks@debian.org>  Tue, 08 May 2007 14:20:50 +0200

usplash (0.4-43-2) unstable; urgency=low

  * initramfs/scripts/init-top/usplash: Enable splash by default on
    installation - patch by David Härdeman. (closes: #397954, #383327)
  * debian/control: Add usplash-theme-debian | usplash-theme depends
    for usplash auto-installed theme.

 -- maximilian attems <maks@debian.org>  Mon, 07 May 2007 18:13:03 +0200

usplash (0.4-43-1) unstable; urgency=low

  [ maximilian attems ]
  Release overlong fun duty

  * new upstream release (svga i386, theming support, input, packaging)
    drop all merged patches plus quilt rules, also the 04-bogl-dpkg-arg.patch,
    as you neeed newer debhelper than sarge version. It no longer leaves
    /var/lib/usplash around.  (closes: 380750, 383249, 397970, 382675)
  * Add lintian override for libusplash.so.0
  * Be vendor neutral in an ubuntu called lsb script. (closes: 384812)
  * debian/usplash.init: Add dep info. (closes: 386349)
    Thanks Petter Reinholdtsen <pere@hungry.com>
  * debian/usplash.init: Only run on stop with splash bootarg.
  * debian/control: Remove dependency on policycoreutils.
  * usplash-theme.h: Add input patch from David.

  [ Otavio Salvador ]
  * Change package to non-native;
  * Revert a usplash-theme.h change that isn't supported on the current
    code;
  * Avoid console-tools and kbd starting when usplash is running;
  * Add myself on uploaders.

 -- Otavio Salvador <otavio@debian.org>  Thu, 15 Mar 2007 15:51:15 -0300

usplash (0.3e) unstable; urgency=low

  * Add 05-usplash_write-quiet.patch from Petter Reinholdtsen
    <pere@hungry.com>. Gets rid of error messages on splash=false boots.

 -- maximilian attems <maks@sternwelten.at>  Fri,  4 Aug 2006 10:04:59 +0200

usplash (0.3d) unstable; urgency=low

  * Add 03-bogl-no-PAGESIZE.patch to the patches/series file to really fix
    the powerpc ftbfs.
  * bogl/Makefile: Ease sarge backports, build with DEB_BUILD_ARCH in
    04-blog-dpkg-arg.patch. Thanks Petter Reinholdtsen <pere@hungry.com>.
    (closes: 380745)
  * debian/usplash.init: Workaround missing ubuntu gdm patches. Call usplash
    in init stop. Thanks Petter Reinholdtsen <pere@hungry.com>.
    (closes: 381009)

 -- maximilian attems <maks@sternwelten.at>  Wed,  2 Aug 2006 09:28:22 +0200


usplash_write.c
/* usplash
 *
 * usplash_write.c - send commands to usplash
 *
 * Copyright © 2006 Canonical Ltd.
 * Copyright © 2005 Matthew Garrett <mjg59@srcf.ucam.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>

#include "usplash.h"


int main(int argc, char *argv[])
{
	int i, fifo_fd;

	if (argc < 2) {
		fprintf(stderr, "Usage: usplash_write COMMAND...\n");
		exit(1);
	}

	if (chdir(USPLASH_DIR) < 0) {
		if (errno != ENOENT)
			perror("chdir");
		exit(0);
	}

	fifo_fd = open(USPLASH_FIFO, O_WRONLY | O_NONBLOCK);
	if (fifo_fd < 0) {
		if ((errno != ENXIO) && (errno != ENOENT))
			perror("open");
		exit(getenv("FAIL_NO_USPLASH") ? 1 : 0);
	}

	for (i = 1; i < argc; i++) {
		size_t argsize;
		ssize_t len;

		/*
		 * POSIX says writes to a fifo larger than PIPE_BUF could
		 * get interleaved, so we must truncate those.
		 */
		argsize = strlen(argv[i]);
		if (argsize + 1 > PIPE_BUF) {
			argsize = PIPE_BUF - 1;
			argv[i][argsize] = '\0';
		}
		
		len = write(fifo_fd, argv[i], argsize + 1);
		if (len < 0) {
			perror("write");
			exit(0);
		}
	}

	return 0;
}
